### ISSUE TITLE
v0.24 Upgrade causing disk space issues on servers

### ISSUE THREAD

**Issue by j2preston** [typesense/typesense#928]
RocksDB garbage collection not clearing disk up after collections marked for deletion.  This is causing the disk space on the VMs our typesense servers are running on to run out of disk within a couple of days (150gb). Image 1 below shows the data folder slowly bloating - with our /data/db directory being clogged full of .sst files (Image 2).

img1
![Capture](https://user-images.githubusercontent.com/47216699/224299277-20798434-7490-4e90-80d9-5e9d97dff3ef.PNG)

img2
![Capture](https://user-images.githubusercontent.com/47216699/224301540-7a349841-2b37-47c8-83ed-8d024958f87a.PNG)

Upgrade servers to new v0.24. Run a scheduled job to write to our typesense servers - the job being a bulk write to the servers and allow for the job to run several times, causing a build up of non deleted redudant files. Image 3 shows Time vs Disk space on the VM when upgrading to v0.24 with scheduled jobs writing to our Typesense server. 

Img3
![Capture](https://user-images.githubusercontent.com/47216699/224310157-5b8b3a6e-01d0-4a3f-a9aa-51c8b04e3f08.PNG)

Expected behaviour is that the RocksDB compaction will kick in on redundant data and permanently remove that data from the disk. As does on v0.23.

Data on the disk continues to ramp up and data directory slowly bloats until it reaches maximum space and bricks the VM. 

Rolling back to v0.23 resolves this issue entirely.

**Typesense Version**:
Typesense v0.24

Worth noting we're running the official typesense docker image.
image: typesense/typesense:0.23.1 - works fine
Image: typesense/typesense:0.24.0 - issue

**Comment by kishorenc** [comment]
@j2preston 

I just tried reproducing this issue but could not. Here's what I did:

1. Imported 1M records
2. Checked the size of the data directory
3. Reimported the same 1M records with `action=upsert` mode, but scrambled up the IDs so that each document is treated as an update
4. Measured the size of the data directory

While the data size grew during the reimport step, eventually compaction kicked in and the eventual file size was back to what it was after step 1.

I have a couple of questions:

1. Would it be possible for you to share a similar disk usage graph for 0.23?
2. Would you be able to try this out on a smaller subset of your data to see if it's reproduceable?

**Comment by kishorenc** [comment]
@j2preston 

Thank you for confirming again. Definitely want to get to the bottom of this. Since I could not reproduce this with a simple collection that I tried with, I need your help in replicating this issue. Would it be possible for you to share the collection schema with me via email? Please email: contact@typesense.org

**Comment by bonsempregos** [comment]
Hi, 

I also have this problem since I enable vector with 384 dimensions. we usually create a new collection every 30 minutes and change alias to the new one, dropping old.

**Comment by kishorenc** [comment]
@bonsempregos 

What's the before and after values of the disk space on restart?

I tried to reproduce this issue locally several times with dummy data but I was not able to find anything amiss. Are you able to reproduce it locally easily?

**Comment by bonsempregos** [comment]
@kishorenc 


df -h /
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1       388G  359G   29G  93% /

VS

df -h /
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1       388G  311G   78G  81% /

maybe somehow is related to https://github.com/typesense/typesense/issues/831 as after running the compactor 

I went from 

 du -sh /var/lib/typesense/*
129G	/var/lib/typesense/db
5.6M	/var/lib/typesense/meta
56G	/var/lib/typesense/state

To

du -sh /var/lib/typesense/*
695M	/var/lib/typesense/db
5.6M	/var/lib/typesense/meta
2.3G	/var/lib/typesense/state

**Comment by kishorenc** [comment]
@bonsempregos

Thanks for confirming that compaction works. I wonder why the automatic level-based compaction isn't kicking in. It does when I tried to repeatedly import + drop a dataset that takes about 350M of disk space. 

What's the size of the `db` directory after a fresh import? Is it also in the range of ~700M?

**Comment by kishorenc** [comment]
@bonsempregos 

I've made some changes to perform automatic compaction during dropping of collections in `typesense/typesense:0.25.0.rc22` -- can you please try this build? NOTE: since 0.25 also contains an upgrade to a newer version of RocksDB, you cannot downgrade by pointing the same data directory back to 0.24.x.

**Comment by kishorenc** [comment]
Yes, [here](https://dl.typesense.org/releases/0.25.0.rc22/typesense-server-0.25.0.rc22-linux-amd64.tar.gz).

**Comment by bonsempregos** [comment]
Hi @kishorenc,

Just tested, dropping 8 collections

Leader:

```
du -sh /var/lib/typesense/* && date
23G	/var/lib/typesense/db
4.8M	/var/lib/typesense/meta
11G	/var/lib/typesense/state
Wed Apr 12 05:31:14 WEST 2023
```

then 

```
du -sh /var/lib/typesense/* && date
3.0G	/var/lib/typesense/db
4.8M	/var/lib/typesense/meta
25G	/var/lib/typesense/state
Wed Apr 12 05:49:14 WEST 2023

```
and finally

```
du -sh /var/lib/typesense/* && date
1.5G	/var/lib/typesense/db
4.8M	/var/lib/typesense/meta
5.0G	/var/lib/typesense/state
Wed Apr 12 06:01:53 WEST 2023
```

All followers are updated too.

Thanks a lot for the fix.

**Comment by kishorenc** [comment]
@bonsempregos That's awesome to hear.

@j2preston Would you be able to test the `0.25.0.rc22` build out to see if it helps?
